src/digraph.rs

```diff
@@ -2426,6 +2426,65 @@ impl PyDiGraph {
         }
     }
 
+    /// Return a new PyDiGraph object for an edge induced subgraph of this graph
+    ///
+    /// The induced subgraph contains each edge in `edges` and each node
+    /// incident to any of those edges.
+    ///
+    /// :param list edges: A list of edge tuples (2-tuples with the source and
+    ///     target node) to generate the subgraph from. In cases of parallel
+    ///     edges for a multigraph all edges between the specified node. In case
+    ///     of an edge specified that doesn't exist in the graph it will be
+    ///     silently ignored.
+    ///
+    /// :returns: The edge subgraph
+    /// :rtype: PyDiGraph
+    ///
+    #[pyo3(text_signature = "(self, edges, /)")]
+    pub fn edge_subgraph(&self, edge_list: Vec<[usize; 2]>) -> PyDiGraph {
+        // Filter non-existent edges
+        let edges: Vec<[usize; 2]> = edge_list
+            .into_iter()
+            .filter(|x| {
+                let source = NodeIndex::new(x[0]);
+                let target = NodeIndex::new(x[1]);
+                self.graph.find_edge(source, target).is_some()
+            })
+            .collect();
+
+        let nodes: HashSet<NodeIndex> = edges
+            .iter()
+            .map(|x| x.iter())
+            .flatten()
+            .copied()
+            .map(NodeIndex::new)
+            .collect();
+        let mut edge_set: HashSet<[NodeIndex; 2]> =
+            HashSet::with_capacity(edges.len());
+        for edge in edges {
+            let source_index = NodeIndex::new(edge[0]);
+            let target_index = NodeIndex::new(edge[1]);
+            edge_set.insert([source_index, target_index]);
+        }
+        let mut out_graph = self.clone();
+        for node in self
+            .graph
+            .node_indices()
+            .filter(|node| !nodes.contains(node))
+        {
+            out_graph.graph.remove_node(node);
+            out_graph.node_removed = true;
```

Comment:
Calling `out_graph.node_removed = true` feels like breaking a unwritten rule, wouldn't it be better to just call `out_graph.remove_node()` on the PyDiGraph class?

Reply:
I just did it this way because to avoid the double conversion of `NodeIndex` -> `usize` -> `NodeIndex`. But I can change this to be:
```suggestion
            out_graph.remove_node(node.index())?;
```
if you'd prefer. I don't think there will be any extra overhead from the double conversion (since a NodeIndex is basically just a thin struct around `usize` anyway (to provide a common trait and debug output for the various unsigned integer types)

Reply:
Double conversion is not ideal, and that would make it a PyResult. I think it is fine to manipulate `.node_removed` then

Reply:
I'm actually not sure why `remove_node` returns a `PyResult`. Looking at the code for the method nothing returns an error. Petgraph's `remove_node` returns an option and `None` if the `NodeIndex` is invalid. But `PyGraph`'s `remove_node` is documented as just ignoring invalid node ids so we probably can remove the pyresult from the method.